src/query/expression/src/schema.rs

```diff
@@ -241,18 +250,306 @@ impl TableSchema {
+            next_column_id: Some(0),
+            column_id_map: Some(BTreeMap::new()),
+            column_id_set: Some(HashSet::new()),
+        }
+    }
+
+    pub fn init_if_need(data_schema: TableSchema) -> Self {
+        // If next_column_id is none, it is an old version needs to compatibility
+        if data_schema.next_column_id.is_none() {
+            Self::new_from(data_schema.fields, data_schema.metadata)
+        } else {
+            data_schema
+        }
+    }
+
+    fn build_from_data_type(
+        data_type: &TableDataType,
+        column_name: &str,
+        next_column_id: &mut u32,
+        column_id_map: &mut BTreeMap<String, u32>,
+        column_id_set: &mut HashSet<u32>,
+    ) {
+        column_id_map.insert(column_name.to_string(), *next_column_id);
+        column_id_set.insert(*next_column_id);
+        match data_type {
+            TableDataType::Tuple {
+                ref fields_name,
+                ref fields_type,
+            } => {
+                for (i, inner_type) in fields_type.iter().enumerate() {
+                    let inner_name = format!("{}:{}", column_name, fields_name[i]);
+                    Self::build_from_data_type(
+                        inner_type,
+                        &inner_name,
+                        next_column_id,
+                        column_id_map,
+                        column_id_set,
+                    );
+                }
+            }
+            TableDataType::Array(a) | TableDataType::Map(a) => {
+                let inner_name = format!("{}:0", column_name);
+                Self::build_from_data_type(
+                    a.as_ref(),
+                    &inner_name,
+                    next_column_id,
+                    column_id_map,
+                    column_id_set,
+                );
+            }
+            TableDataType::Nullable(a) => {
+                Self::build_from_data_type(
+                    a.as_ref(),
+                    column_name,
+                    next_column_id,
+                    column_id_map,
+                    column_id_set,
+                );
+            }
+            _ => {
+                *next_column_id += 1;
+            }
         }
     }
 
+    fn build_members_from_fields(
+        fields: &[TableField],
+        column_id_map: Option<BTreeMap<String, u32>>,
+        next_column_id_opt: Option<u32>,
+    ) -> (u32, BTreeMap<String, u32>, HashSet<u32>) {
+        let mut next_column_id = 0;
+        let mut has_column_id_map_inited = false;
+        let mut column_id_map = match column_id_map {
+            Some(column_id_map) => {
+                has_column_id_map_inited = !column_id_map.is_empty();
+                column_id_map
+            }
+            None => BTreeMap::new(),
+        };
+        let has_next_column_id_inited = match next_column_id_opt {
+            Some(next_column_id) => next_column_id > 0,
+            None => false,
+        };
```

Comment:
`next_column_id_opt.unwrap_or_default()` would be quite enough. You do not really need to care about if it is initialized, IMHO.